woltka/tests/test_workflow.py

```diff
@@ -136,35 +167,211 @@ def test_parse_samples(self):
         self.assertEqual(str(ctx.exception), (
             '"im/not/path" is not a valid file or directory.'))
 
-    def test_workflow(self):
-        # simplest gotu workflow
-        input_path = join(self.datdir, 'align', 'bowtie2')
-        output_path = join(self.tmpdir, 'tmp.tsv')
-        obs = workflow(input_path, output_path)['none']
-        self.assertEqual(obs['S01']['G000011545'], 48)
-        self.assertNotIn('G000007145', obs['S02'])
-        self.assertEqual(obs['S03']['G000009345'], 640)
-        with open(output_path, 'r') as f:
+    def test_build_mapper(self):
+        # plain
+        obs = build_mapper()
+        self.assertEqual(obs.__class__.__name__, 'Plain')
+
+        # ordinal
+        fp = join(self.tmpdir, 'coords.txt')
+        with open(fp, 'w') as f:
+            f.write('>G1\n1\t10\t20\n2\t35\t50\n')
+        obs = build_mapper(fp)
+        self.assertEqual(obs.__class__.__name__, 'Ordinal')
+
+        # ordinal with overlap threshold
+        obs = build_mapper(fp, 75)
+        self.assertEqual(obs.__class__.__name__, 'Ordinal')
+        self.assertEqual(obs.th, 0.75)
+        remove(fp)
+
+    def test_prepare_ranks(self):
+        # no rank
+        obs = prepare_ranks()
+        self.assertListEqual(obs[0], ['none'])
+
+        # single rank
+        obs = prepare_ranks('genus')
+        self.assertListEqual(obs[0], ['genus'])
+
+        # multiple ranks
+        obs = prepare_ranks('phylum,genus,species')
+        self.assertListEqual(obs[0], ['phylum', 'genus', 'species'])
+
+        # with output read map directory
+        obs = prepare_ranks(outmap_dir=self.tmpdir)
+        self.assertTupleEqual(obs, (['none'], {'none': self.tmpdir}))
+
+        # make multiple output read map directories
+        obs = prepare_ranks('ko,mo', self.tmpdir)
+        self.assertListEqual(obs[0], ['ko', 'mo'])
+        self.assertDictEqual(obs[1], {'ko': join(self.tmpdir, 'ko'),
+                                      'mo': join(self.tmpdir, 'mo')})
+        rmtree(join(self.tmpdir, 'ko'))
+        rmtree(join(self.tmpdir, 'mo'))
+
+    def test_build_hierarchy(self):
```

Comment:
Can you break this up into mutliple tests?
You want to be able to tell if some change breaks each functionality individually

Reply:
Each functionality (e.g., `read_names`) is tested with more detail in another script ([test_tree.py](https://github.com/qiyunzhu/woltka/blob/master/woltka/tests/test_tree.py)). Here `build_hierarchy` is but a workflow calling individual functions. So I tend to make the current test simple as whether it can trigger individual functions and get desired result. How does this sound?

Reply:
While the opinion may be extreme, I tend to think "one assert statement per test". If someone who is not familiar with your code base is working on it, it greatly enhances their workflow if they can see all breaking tests at once. It really stinks to fix an issue in line 218 to then just run into another issue on line 226 that you didn't know was there when you run the tests the original time.